---
title: "Motif analysis with Signac"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

In this tutorial, we will perform DNA sequence motif analysis in Signac. We will explore two complementary options for performing motif analysis: one by finding overrepresented motifs in a set of differentially accessible peaks, one method performing differential motif activity analysis between groups of cells.

In this demonstration we use data from the adult mouse brain. See our [vignette](https://satijalab.org/signac/articles/mouse_brain_vignette.html) for the code used to generate this object, and links to the raw data. First, load the required packages and the pre-computed Seurat object:

```{r message=FALSE, warning=FALSE}
library(Signac)
library(Seurat)
library(JASPAR2018)
library(TFBSTools)
library(BSgenome.Mmusculus.UCSC.mm10)
library(patchwork)
set.seed(1234)
```


```{r message=FALSE, warning=FALSE}
mouse_brain <- readRDS("../vignette_data/adult_mouse_brain.rds")
mouse_brain
```

```{r message=FALSE, warning=FALSE}
p1 <- DimPlot(mouse_brain, label = TRUE, pt.size = 0.1) + NoLegend()
p1
```

## The Motif class

To facilitate motif analysis in Signac, we have create the `Motif` class to store all the required information, including a list of position weight matrices (PWMs) or position frequency matrices (PFMs) and a motif occurrence matrix. Here, we construct a `Motif` object and add it to our mouse brain dataset. A motif object can be added to any Seurat assay using the `AddMotifObject` function:

```{r message=FALSE, warning=FALSE}
# Get a list of motif position frequency matrices from the JASPAR database
pfm <- getMatrixSet(
  x = JASPAR2018,
  opts = list(species = 9606, all_versions = FALSE)
)

# Scan the DNA sequence of each peak for the presence of each motif
motif.matrix <- CreateMotifMatrix(
  features = StringToGRanges(rownames(mouse_brain), sep = c(":", "-")),
  pwm = pfm,
  genome = 'mm10',
  sep = c(":", "-"),
  use.counts = FALSE
)

# Create a new Mofif object to store the results
motif <- CreateMotifObject(
  data = motif.matrix,
  pwm = pfm
)

# Add the Motif object to the assay
mouse_brain[['peaks']] <- AddMotifObject(
  object = mouse_brain[['peaks']],
  motif.object = motif
)
```

In order to test for overrepresented motifs, we also need to compute some sequence characteristics of the peaks, such as GC content, sequence length, and dinucleotide frequency.
The `RegionStats` function computes this information for us and stores the results in the feature metadata in the Seurat object.

```{r message=FALSE, warning=FALSE}
mouse_brain <- RegionStats(
  object = mouse_brain,
  genome = BSgenome.Mmusculus.UCSC.mm10,
  sep = c(":", "-")
)
```

## Finding overrepresented motifs

To identify potentially important cell-type-specific regulatory sequences, we can search for DNA motifs that are overrepresented in a set of peaks that are differentially accessible between cell types.

Here, we find differentially accessible peaks between Pvalb and Sst inhibitory interneurons. We then perform a hypergeometric test to test the probability of observing the motif at 
the given frequency by chance, comparing with a background set of peaks matched for GC content. 

```{r message=FALSE, warning=FALSE}
da_peaks <- FindMarkers(
  object = mouse_brain,
  ident.1 = 'Pvalb',
  ident.2 = 'Sst',
  only.pos = TRUE,
  test.use = 'LR',
  latent.vars = 'nCount_peaks'
)

# Test the differentially accessible peaks for overrepresented motifs
enriched.motifs <- FindMotifs(
  object = mouse_brain,
  features = head(rownames(da_peaks), 1000)
)
# sort by p-value and fold change
enriched.motifs <- enriched.motifs[order(enriched.motifs[, 7], -enriched.motifs[, 6]), ]
```

```{r}
head(enriched.motifs)
```

We can also plot the position weight matrices for the motifs, so we can visualize the different motif sequences.

```{r}
MotifPlot(
  object = mouse_brain,
  motifs = head(rownames(enriched.motifs))
)
```

We and others have previously shown that Mef-family motifs, particularly *Mef2c*, are enriched in Pvalb-specific peaks in scATAC-seq data
(https://doi.org/10.1016/j.cell.2019.05.031; https://doi.org/10.1101/615179),
and further shown that *Mef2c* is required for the development of Pvalb interneurons (https://www.nature.com/articles/nature25999).
Here our results are consistent with these findings, and we observe a strong enrichment of Mef-family motifs in the top results from `FindMotifs`.

## Computing motif activities

We can also compute a per-cell motif activity score by running [chromVAR](https://greenleaflab.github.io/chromVAR/index.html). This allows us to visualize motif activities per cell, and also provides an alternative method of identifying differentially-active motifs between cell types.

ChromVAR identifies motifs associated with variability in chromatin accessibility between cells.
See the chromVAR [paper](https://www.nature.com/articles/nmeth.4401) for a complete description of the method.

```{r message=FALSE, warning=FALSE, fig.width=12}
mouse_brain <- RunChromVAR(
  object = mouse_brain,
  genome = BSgenome.Mmusculus.UCSC.mm10
)

DefaultAssay(mouse_brain) <- 'chromvar'

# look at the activity of Mef2c, the top hit from the overrepresentation testing
p2 <- FeaturePlot(
  object = mouse_brain,
  features = rownames(enriched.motifs)[[1]],
  min.cutoff = 'q10',
  max.cutoff = 'q90',
  pt.size = 0.1
)
p1 + p2
```

We can also directly test for differential activity scores between cell types. This tends to give similar results as performing an enrichment test on differentially accessible peaks between the cell types (shown above)

```{r message=FALSE, warning=FALSE}
differential.activity <- FindMarkers(
  object = mouse_brain,
  ident.1 = 'Pvalb',
  ident.2 = 'Sst',
  only.pos = TRUE,
  test.use = 'LR',
  latent.vars = 'nCount_peaks'
)

MotifPlot(
  object = mouse_brain,
  motifs = head(rownames(differential.activity)),
  assay = 'peaks'
)
```

<details>
  <summary>**Session Info**</summary>
  
```{r}
sessionInfo()
```

</details>

